office/ProjectLibre/Portfile

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:filetype=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           java 1.0
+
+# retrieving shasum 858bc0d6f82fdea6c67f145c768e70b48beb1b16 from master because there is no tag 1.8.0 set
```

Comment:
Do you actually need a specific git commit?

I'm just asking because using the released files from https://sourceforge.net/projects/projectlibre/files/ProjectLibre/1.8/ generally sounds like a more useful idea. Plus, I had the impression that one can only download a specific shasum from sourceforge if it was still in cache. After a while I was no longer able to get the files (I don't claim I downloaded the same type of the files, and it was already some time ago).

I almost never manage to get it right the first time, but it must be something similar to
```
master_sites sourceforge:project/projectlibre/ProjectLibre/1.8
```
with some additional trickery to automatically extract 1.8 from 1.8.0.

Reply:
Indeed, it should be better to download from a release tag. 
I did several tests, but I never managed to get to the **source** package (which contains the ant build system that create the .app bundle).
In particular, when I try this settings (basically the link that you propose):
```
master_sites        https://sourceforge.net/projects/projectlibre/files/ProjectLibre/1.8/
distfiles           projectlibre-1.8.0.tar.gz
```
I get only the .jar file (plus some others).
When I try these settings:
```
master_sites        sourceforge:projects/projectlibre/files/ProjectLibre/1.8/download
distfiles           projectlibre-1.8.0.tar.gz
```
I'm redirected to a stub html page from the mirror and I get the following error:
```
The non-matching file appears to be HTML.
```
What should I do? Please advise.

Reply:
Which file do you need to download manually to get the desired contents? I guess it's one of those from https://sourceforge.net/projects/projectlibre/files/ProjectLibre/1.8/

Reply:
To my understanding the link you give contains only binary packages and no source tarball. I dont'k know MacPorts policy for this case, but if you still want to download from that link I suppose the easyest way would be to fetch the dmg and retrieve the .app inside or to download the .jar and bundle in a .app (I'm not able to do any of these two straight away and I'd need further investigation though). Then I may package the sources in a sub port.

Reply:
No, we don't want to fetch the `dmg`. But there are various other options:
* `projectlibre.jar`
* `projectlibre-1.8.0.tar.gz`
* `projectlibre-1.8.0.zip`
which I didn't inspect, but they look like potential candidates to me.

Reply:
Let me check those files ...

Reply:
> No, we don't want to fetch the `dmg`. But there are various other options:
> 
>     * `projectlibre.jar`
This a binary package (opcode compiled with javac)

>     * `projectlibre-1.8.0.tar.gz`
This contains the .jar (and some few other things)
 
>     * `projectlibre-1.8.0.zip`
This again contains the .jar (and some few other things)

Reply:
I checked now and I admit that it's super weird.

I tried to download
```
wget https://freefr.dl.sourceforge.net/code-snapshots/git/p/pr/projectlibre/code.git/projectlibre-code-858bc0d6f82fdea6c67f145c768e70b48beb1b16.zip
```
and it fails with "404 Not Found", which is kind of consistent with my past experience. I'm also not able to fetch the sources from your Portfile. Can you please try to delete (do the backup first) `/opt/local/var/macports/distfiles/ProjectLibre` and try to fetch the port again? From my past experience SourceForge is able to make the tarball, but only if you click the button on the website. Then it keeps the file in cache for a while, but it disappears soon after. If this worked, we would have probably written a PortGroup to support fetching the files this way, in the same way as it is done for GitHub.

I would primarily ask upstream why they don't provide the sources elsewhere, but I'm afraid that the only way to support this properly then is to fetch from git, unless you find a better way, or at least I hope I'm wrong. Using git has its own set of drawbacks, the main one being that the sources are neither stored on the disk, nor on our mirror (at least not until we finish and merge the `vcs-fetch` branch), so you need to download everything again and again in case you try to build multiple times.

Again, I would try to upstream. Even if upstream doesn't answer, we could use git. An alternative would be to make sure that the file is cached on sourceforge just before merging this. Then we would get that file mirrored, and users could fetch it from our mirrors. Again it would not be trivial to update, but well ...

Reply:
> I tried to download
> 
> ```
> wget https://freefr.dl.sourceforge.net/code-snapshots/git/p/pr/projectlibre/code.git/projectlibre-code-858bc0d6f82fdea6c67f145c768e70b48beb1b16.zip
> ```
> 
> and it fails with "404 Not Found", which is kind of consistent with my past experience. I'm also not able to fetch the sources from your Portfile. Can you please try to delete (do the backup first) `/opt/local/var/macports/distfiles/ProjectLibre` and try to fetch the port again? From my past experience SourceForge is able to make the tarball, but only if you click the button on the website. Then it keeps the file in cache for a while, but it disappears soon after. If this worked, we would have probably written a PortGroup to support fetching the files this way, in the same way as it is done for GitHub.

I couldn't believe this caching behaviour, but you're right: the Portfile doesn't work as it is.

> I would primarily ask upstream why they don't provide the sources elsewhere, but I'm afraid that the only way to support this properly then is to fetch from git, unless you find a better way, or at least I hope I'm wrong. 

I opened a ticket to upstream on sourceforge. I must say they don't look very responsive...
I saw on GitHub this repository https://github.com/claur/ProjectLibre which looks the same project from the same author (I asked for confirmation to upstream). 
**What about fetching from there?**
 
> Again, I would try to upstream. Even if upstream doesn't answer, we could use git. An alternative would be to make sure that the file is cached on sourceforge just before merging this. Then we would get that file mirrored, and users could fetch it from our mirrors. Again it would not be trivial to update, but well ...

I don't understand completely what/if I should do.

> Yet another alternative (but also strange) would be to reproduce their build which seems relatively straightforward (mostly just copying a few files around). Or the few text files could be copied to MacPorts. But that's all very subopmital, and I'm not a huge fan of any of those solutions.

Do you mean bundling a .app from the jar? I started to investigate ```appbundler``` and ```ant``` and ```build.xml```, but couldn't find a solution yet.

Reply:
> > I would primarily ask upstream why they don't provide the sources elsewhere, but I'm afraid that the only way to support this properly then is to fetch from git, unless you find a better way, or at least I hope I'm wrong.
> 
> I opened a ticket to upstream on sourceforge. I must say they don't look very responsive...

Last commit 9 months ago? That in fact doesn't really look super promising.

> I saw on GitHub this repository https://github.com/claur/ProjectLibre which looks the same project from the same author (I asked for confirmation to upstream).
> **What about fetching from there?**

Not ideal, but I guess than under these circumstances I would accept it. This seems like a private clone that's probably manually synced, without even syncing the few tags which do exist.

> > Again, I would try to upstream. Even if upstream doesn't answer, we could use git. An alternative would be to make sure that the file is cached on sourceforge just before merging this. Then we would get that file mirrored, and users could fetch it from our mirrors. Again it would not be trivial to update, but well ...
> 
> I don't understand completely what/if I should do.

If we went that way, then nothing. Just try to download the file immediately before someone would merge it, and then we would have the file mirrored in MacPorts.

> > Yet another alternative (but also strange) would be to reproduce their build which seems relatively straightforward (mostly just copying a few files around). Or the few text files could be copied to MacPorts. But that's all very subopmital, and I'm not a huge fan of any of those solutions.
> 
> Do you mean bundling a .app from the jar? I started to investigate `appbundler` and `ant` and `build.xml`, but couldn't find a solution yet.

I wouldn't spend too much time on this. It was just a random brainstorming idea when I ran out of reasonable ways. What I meant was that apart from some huge files, the `openproj_build` seems to contain a very small number of files, probably all of them really small. Under mac there's just:
```
./resources/mac/JavaAppLauncher # only 64-bit, no clue what the requirements are
./resources/mac/old/Info.plist
./resources/mac/embeded/Info.plist
./resources/mac/projectlibre.icns
./resources/mac/new/Info.plist
./resources/mac/new/PkgInfo
./resources/mac/JavaApplicationStub # just ppc and i386
./resources/mac/PkgInfo
```
and it might be that just copying these files over could get the job done. But I'm not really sure I would want to go that route and overcomplicate the life.

Maybe go with GitHub mirror for now?

Reply:
Ok to use the GitHub repository for now.
I made a mistake with my branch and used a force push to resolve, I hope everything is all right.